purchases/src/main/kotlin/com/revenuecat/purchases/blockstore/BlockstoreHelper.kt

```diff
@@ -0,0 +1,140 @@
+package com.revenuecat.purchases.blockstore
+
+import android.content.Context
+import com.google.android.gms.auth.blockstore.Blockstore
+import com.google.android.gms.auth.blockstore.BlockstoreClient
+import com.google.android.gms.auth.blockstore.DeleteBytesRequest
+import com.google.android.gms.auth.blockstore.RetrieveBytesRequest
+import com.google.android.gms.auth.blockstore.RetrieveBytesResponse.BlockstoreData
+import com.google.android.gms.auth.blockstore.StoreBytesData
+import com.revenuecat.purchases.CustomerInfo
+import com.revenuecat.purchases.PurchasesException
+import com.revenuecat.purchases.common.debugLog
+import com.revenuecat.purchases.common.errorLog
+import com.revenuecat.purchases.identity.IdentityManager
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.launch
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+internal class BlockstoreHelper(
+    applicationContext: Context,
+    private val identityManager: IdentityManager,
+    private val blockstoreClient: BlockstoreClient = Blockstore.getClient(applicationContext),
+    private val ioScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
+    private val mainScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main),
+) {
+
+    private companion object {
+        const val BLOCKSTORE_USER_ID_KEY = "revenuecat_blockstore_user_id"
+        const val BLOCKSTORE_MAX_ENTRIES = 16
+    }
+
+    fun storeUserIdIfNeeded(customerInfo: CustomerInfo) {
+        if (!identityManager.currentUserIsAnonymous() || customerInfo.allPurchasedProductIds.isEmpty()) return
+        ioScope.launch {
+            val blockstoreData = try {
+                getBlockstoreData()
+            } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
+                errorLog { "Failed to retrieve Block store data. Will not store userId. Error: ${e.message}" }
+                return@launch
+            }
+            val currentUserId = identityManager.currentAppUserID
+            if (shouldStoreBlockstoreUserId(blockstoreData)) {
+                try {
+                    storeBlockstoreUserId(currentUserId)
+                } catch (@Suppress("TooGenericExceptionCaught") e: Exception) {
+                    errorLog { "Failed to store user Id in Block store: $e" }
+                }
+            }
+        }
+    }
+
+    fun recoverAndAliasBlockstoreUserIfNeeded(callback: () -> Unit) {
```

Comment:
Naming suggestion:

```suggestion
    fun aliasCurrentAndStoredUserIdsIfNeeded(callback: () -> Unit) {
```

I think the "recover" part is beyond the concern of the `BlockstoreHelper`. It just deals with user IDs. It doesn't really care about the side effect of purchases being recovered.